risc0/zkvm/src/dev_mode.rs

```diff
@@ -0,0 +1,184 @@
+    ///
+    /// NOTE: This is thread-local because it entering a DevModeContext in one thread should not
+    /// effect other threads.
+    static DEV_MODE_CONTEXT_DEPTH: RefCell<u32> = RefCell::new(0);
+}
+
+// PhantomData<*cont ()> prevents Rust from automatically implementing Send. This is important
+// because we don't want this to be held across an `.await`.
+/// When a [DevModeContext] exists on a the stack, dev mode will be enabled.
+pub struct DevModeContext(PhantomData<*const ()>);
+
+impl DevModeContext {
+    /// Enter a dev mode context. While this value is held, dev mode will be enabled.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # if cfg!(feature = "disable-dev-mode") { return; }
+    /// # use risc0_zkvm::{DevModeContext, is_dev_mode};
+    /// {
+    ///     let _dev_mode_ctx = DevModeContext::enter();
+    ///     assert_eq!(true, is_dev_mode());
+    /// }
+    ///
+    /// // Enter dev mode based on a runtime condition.
+    /// {
+    ///     let enter_dev_mode: bool = rand::random();
+    ///     let _dev_mode_ctx = enter_dev_mode.then(DevModeContext::enter);
+    ///     assert_eq!(enter_dev_mode, is_dev_mode());
+    /// }
+    /// ```
+    #[must_use]
+    pub fn enter() -> Self {
+        if cfg!(feature = "disable-dev-mode") {
+            panic!("DevModeContext::enter() called when disable-dev-mode feature is enabled. These are inconsistent configurations.");
+        }
+        DEV_MODE_CONTEXT_DEPTH.with_borrow_mut(|x| *x += 1);
+        Self(PhantomData)
+    }
+
+    /// Exit a dev mode context.
+    ///
+    /// When the last [DevModeContext] is exited, dev mode will be disabled. Exiting the context
+    /// happens automatically on drop, so you should not need to call this function manually unless
+    /// you want to explicitly end the context early.
+    pub fn exit(self) {
+        drop(self);
+    }
+}
+
+impl Drop for DevModeContext {
+    /// Exit the dev mode context on drop.
+    fn drop(&mut self) {
+        DEV_MODE_CONTEXT_DEPTH.with_borrow_mut(|x| {
+            *x = x
+                .checked_sub(1)
+                .expect("underflow of DEV_MODE_CONTEXT_DEPTH occurred")
+        });
+    }
+}
+
+/// Returns `true` if dev mode is enabled.
+///
+/// When dev mode is enabled, [FakeReceipt][crate::FakeReceipt] can be used to test applications.
+/// In particular, the verifier will accept fake receipts as a valid proof of
+/// execution, even though it provides no cryptographic guarantees. Additionally, the prover will
+/// produce a fake receipt instead of spending computation resources to produce a full receipt.
+///
+/// Dev mode can be enabled in one of two ways:
+/// 1. By setting the `RISC0_DEV_MODE` environment variable to "1", "true", or "yes". This does not
+///    work when building for the guest (i.e. `target_os = zkvm`).
+/// 2. By calling [DevModeContext::enter()] and holding the returned value.
```

Comment:
What are some use cases for `DevModeContext::enter`, especially where `RISC0_DEV_MODE` falls short?